src/bsoncxx/types/bson_value/value.hpp

```diff
@@ -51,6 +51,9 @@ class BSONCXX_API value {
     ///
     explicit value(const view&);
 
+    value(b_bool b);  // TODO: use BSONCXX_ENUM for all b_ types
```

Comment:
As the ticket describes, the b_types are not things we really want to encourage users to use.  There's a few reasons it's better to hide these types, the biggest reason being that they mostly do not own their own memory.

For now, let's only provide the constructors from the base types themselves--so, here, provide only the constructor that takes a `bool` and not the one that takes `types::b_bool`.

Unfortunately, this probably means you can't use the slick `BSONCXX_ENUM` :( though you could probably do something fancy to strip the `b_` prefixes off, I'm not sure it's worth it.

Reply:
How would we create values that have the same underlying type? For example, is `value(std::string str)` a utf8, code or symbol.

Reply:
> As the ticket describes, the b_types are not things we really want to encourage users to use.

I generally agree that using the b_types structs is potentially dangerous because it is not clear that it does not own its data. I'm not sure what the long term goal would be to remove them (in favor of view/value types that clarify ownership?).

> How would we create values that have the same underlying type? For example, is value(std::string str) a utf8, code or symbol?

Good point. If we do want to avoid using the `types` structs in the API, we'd need to avoid ambiguity when there is the same underlying type.
If we're sure we want to avoid the b_types structs, I'd be in favor of explicit named static constructors, like `value::from_string(const std::string& str)` and `value::from_symbol (const std::string& str)`. That seems straightforward at the cost of some extra verbosity.

Reply:
Sort of unrelated, but I think the Genny users also complained about our aversion to allocating memory:

> The library seems to go out of its way to avoid allocating memory. While that is admirable, it does not make it easy when you are building essentially tree-like structures (such as BSON), and it hardly matters when it must eventually be sent over the network anyway.

This could also be circumvented by just having a constructor like:
```
template <typename... Args>
value(type k_type = k_undefined, Args&&... args);
```

Reply:
> Until we remove the b_types we should probably support conversions between a generic value and a specific value.

That is a fair point. Unless we have a complete solution of how to replace `b_types` structs, we'd have an asymmetric API, where users must use the `b_type` structs when accessing element BSON values, but cannot use them to construct BSON values.

E.g. currently users get `b_type` structs from iterating, and can use them to construct document values:
```
bsoncxx::types::b_regex regex = doc.view()["field"].get_regex();
bsoncxx::document::value doc = make_document(kvp("regex", regex));
```

But we'd be requiring users to split out the fields the `b_type` in order to construct a `bson_value::value`
```
bsoncxx::types::bson_value val {regex.regex, regex.options};
```

So, in chatting with Clyde, the proposal is that the constructor should support the same set of values that the `bsoncxx::builder::core:append()` accepts. @samantharitter do you agree?

Reply:
Hm, the type ambiguity does make this more complicated, I hadn't thought of that.  I like both @kevinAlbs 's factory method solution and @bazile-clyde 's type argument method solution as workarounds for that.

I agree that omitting constructors that take the b_types would lead to an asymmetrical API.  But, I'm also not convinced that that alone is a reason for us to add them to the driver.  There are plenty of quirks in the current driver API, and this particular feature is one that I requested, it didn't come from the community.

I think it would help if we figured out what our ideal future BSON library looks like and work backwards from that design.  For example, if we would like to ultimately remove the `b_type`s from the public API, it doesn't make sense to introduce more methods that use them, as this will lead to more user code that ultimately needs to be changed when we deprecate and then remove those types.

Reply:
Per discussion, let's keep the `b_type` constructors to give a consistent API. If we remove the `b_type` structs in a future design, it will be part of a larger change.

Reply:
@samantharitter thinking about it a bit more, the only way I can see us removing the `b_types` is if we create a class for each of the BSON types that don't easily map to built-in C++ types. This would be similar to what's done for `decimal128` and `oid`. Of course, these new types would be owning. We'd then only need a class for each BSON type and this class... I _think_. 

I think the next steps though would be a design doc of some kind in the way future but wanted to run this idea by both of you.